### ISSUE TITLE
`!` in JSON.renderYAML breaks some parsers

### ISSUE THREAD

**Issue by timbertson** [dhall-lang/dhall-lang#931]
Technically the parsers are noncompliant, but I'm trying to use `dhall to-directory-tree` so I can only use `renderYAML` instead of the nicer `dhall-yaml`.

The current rendering renders all keys as e.g. `! "jobs":`. Which seems to be valid YAML, but some parsers (Github Workflows is the one I'm trying to appease) fail to parse this. The error is:

> Line: 1, Col: 1, Idx: 0) - (Line: 1, Col: 2, Idx: 1): While parsing a tag, did not find expected tag URI

I don't know what library they're using under the hood (maybe YamlDotNet given the similarity to https://stackoverflow.com/questions/9664113/what-does-a-single-exclamation-mark-do-in-yaml).

So can I ask what the intent is with these `!` prefixes, and what the tradeoff would be for dropping them? (or more straightforwardly: can we just drop them?)

I assume it might result in a less strict interpretation of the key type, but I'm struggling to see how a lack of strictness in keys (which are always double-quoted strings) would be that problematic in practice.

**Comment by sjakobi** [comment]
@matheus23 Could you say what the motivation for using the exclamation marks was?

**Comment by Gabriel439** [comment]
@timbertson: One reason for supporting YAML is for unit tests (e.g. for `dhall-kubernetes`), so you can write something like:

```
let test =
      :   renderYAML something
        ≡ ''
          foo: bar
          …
          ''
```

... which makes the test much more readable for large YAML values

I also did some research on the exclamation marks and here is what I learned:

* They were introduced in YAML specification version 1.1

  … so the parser you are using might have been written for an older version

* They are used to disable [tag resolution](https://yaml.org/spec/1.2/spec.html#tag/resolution/)

  Specifically, "!" is a way to force the YAML interpreter to treat something as either a list, map, or string, and not some custom user tag.

In other words, the `!` is being used to force the interpreter to treat the keys of a map as strings.

That said, I *think* it should be fine to omit the exclamation marks, since I don't think many users will run into tag resolution issues in practice.  Also, we haven't been consistent in our use of `!`.  For example, we haven't been using it to similarly disable tag resolution for sequences, maps, or non-key strings.

**Comment by matheus23** [comment]
> @matheus23 Could you say what the motivation for using the exclamation marks was?

I guess the right move would have been to document this. The only comment on this decision I found was this:

> In the worst case (multiline strings with trailing whitespace), libyaml simply resorts to using single-line double-quotation as an escaping mechanism:
> ...
> For everything in between, there are many ways to handle it: Various multiline formats that preserve or don't preserve whitespace and single-quoted strings with an exclamation mark.

I guess I copied this behavior over from what I've seen `libyaml` output to look like.

**Comment by timbertson** [comment]
> force the interpreter to treat the keys of a map as strings

I guess I don't understand when that wouldn't be the case, if a yaml parser sees `"..."`. I can see the need with plain scalars (like `no`, `true`, `1` etc which could be ambiguously a string _or_ a bool/number), but a double-quoted string seems unambiguous already.

**Comment by Gabriel439** [comment]
@timbertson: I think it's because an implementation can attach semantic information to things that would otherwise be unambiguous (e.g. treating a string as a date and validating that it matches an expected format)

However, I don't really see anybody arguing strongly for keeping the `!`s, so I will put up a pull request to remove them

**Comment by Gabriel439** [comment]
Actually, I'll probably wait for https://github.com/dhall-lang/dhall-lang/pull/929 to merge first to avoid unnecessary merge conflicts

**Comment by timbertson** [comment]
I'm still curious what that would actually look like (i.e. what would the code look like to tell a yaml parser to behave in this way). It _seems_ like it would have to be pretty wild but maybe I just lack imagination.

But it doesn't really matter, if we're happy to drop the `!` then I'm happy. Just curious about how weird YAML can get ;). No worries on waiting for #929 since that also benefits me 😀

**Comment by Gabriel439** [comment]
The change is up here: https://github.com/dhall-lang/dhall-lang/pull/934
